tsdb/engine/tsm1/bool.go

```diff
@@ -0,0 +1,136 @@
+package tsm1
+
+// bool encoding uses 1 bit per value.  Each compressed byte slice contains a 1 byte header
+// indicating the compression type, followed by a variable byte encoded length indicating
+// how many booleans are packed in the slice.  The remaining bytes contains 1 byte for every
+// 8 boolean values encoded.
+
+import "encoding/binary"
+
+const (
+	// boolUncompressed is an uncompressed boolean format
+	boolUncompressed = 0
+	// boolCompressedBitPacked is an bit packed format using 1 bit per boolean
+	boolCompressedBitPacked = 1
+)
+
+type BoolEncoder interface {
+	Write(b bool)
+	Bytes() ([]byte, error)
+}
+
+type BoolDecoder interface {
+	Next() bool
+	Read() bool
+	Error() error
+}
+
+type boolEncoder struct {
+	// The encoded bytes
+	bytes []byte
+
+	// The current byte being encoded
+	b byte
+
+	// The number of bools packed into b
+	i int
+
+	// The total number of bools written
+	n int
+}
+
+func NewBoolEncoder() BoolEncoder {
+	return &boolEncoder{}
+}
+
+func (e *boolEncoder) Write(b bool) {
+	// If we have filled the current byte, flush it
+	if e.i >= 8 {
```

Comment:
Let me put it another way. If this value should never be greater than 8,
then we should holler loudly if it is, and not just keep going. Since this
is such a core part of the system, we shouldn't ignore any sign it is not
operating the way we expect.

If it can be greater than 8 under normal operation then we are fine
obviously.

On Monday, October 5, 2015, Philip O'Toole <philip@influxdb.com> wrote:

> Sure, but if bits are written one at a time, why >=? It should just be ==
> right?
>
> On Monday, October 5, 2015, Jason Wilder <notifications@github.com
> <javascript:_e(%7B%7D,'cvml','notifications@github.com');>> wrote:
>
>> In tsdb/engine/tsm1/bool.go
>> <https://github.com/influxdb/influxdb/pull/4308#discussion_r41167948>:
>>
>> > +	b byte
>> > +
>> > +	// The number of bools packed into b
>> > +	i int
>> > +
>> > +	// The total number of bools written
>> > +	n int
>> > +}
>> > +
>> > +func NewBoolEncoder() BoolEncoder {
>> > +	return &boolEncoder{}
>> > +}
>> > +
>> > +func (e *boolEncoder) Write(b bool) {
>> > +	// If we have filled the current byte, flush it
>> > +	if e.i >= 8 {
>>
>> That is the counter of how many booleans have been written for the
>> current byte. When the ninth boolean is written we need to flush the
>> current byte and start over for the new byte.
>>
>> —
>> Reply to this email directly or view it on GitHub
>> <https://github.com/influxdb/influxdb/pull/4308/files#r41167948>.
>>
>